File: backend/routes/recentlyPlayed.js
const express = require('express');
const router = express.Router();
const auth = require('../middleware/auth');
const { query } = require('../db/postgres');

// Get recently played games for the logged-in user
router.get('/', auth, async (req, res) => {
    try {
        const userId = parseInt(req.user.id, 10);
        const result = await query('SELECT recently_played FROM users WHERE id = $1', [userId]);

        if (result.rows.length === 0) {
            return res.status(404).json({ message: 'User not found' });
        }

        const recent = result.rows[0].recently_played || [];
        res.json(recent);
    } catch (error) {
        console.error('Error fetching recently played games:', error);
        res.status(500).json({ message: 'Server error loading recently played' });
    }
});

// Add a game to recently played
router.post('/track', auth, async (req, res) => {
    try {
        const userId = parseInt(req.user.id, 10);
        const { gameId, gameName } = req.body;

        if (!gameId || !gameName) {
            return res.status(400).json({ message: 'Game ID and name are required' });
        }

        const userResult = await query('SELECT recently_played FROM users WHERE id = $1', [userId]);
        if (userResult.rows.length === 0) {
            return res.status(404).json({ message: 'User not found' });
        }

        let recent = userResult.rows[0].recently_played || [];
        // Filter out existing occurrence of this game
        recent = recent.filter(g => g.gameId !== gameId);
        // Prepend new play event
        recent.unshift({
            gameId,
            gameName,
            lastPlayed: new Date().toISOString()
        });
        // Keep top 5
        recent = recent.slice(0, 5);

        await query('UPDATE users SET recently_played = $1 WHERE id = $2', [JSON.stringify(recent), userId]);

        res.json(recent);
    } catch (error) {
        console.error('Error tracking recently played game:', error);
        res.status(500).json({ message: 'Server error saving recently played' });
    }
});

module.exports = router;

File: backend/internal/handlers/recently_played.go
package handlers

import (
	"context"
	"encoding/json"
	"net/http"
	"time"

	"game-on-backend/internal/middleware"
	"game-on-backend/internal/models"

	"github.com/jackc/pgx/v5/pgxpool"
)

type RecentlyPlayedHandler struct {
	pool *pgxpool.Pool
}

func NewRecentlyPlayedHandler(pool *pgxpool.Pool) *RecentlyPlayedHandler {
	return &RecentlyPlayedHandler{pool: pool}
}

// GetRecentlyPlayed retrieves the player's recent game history from PostgreSQL
func (h *RecentlyPlayedHandler) GetRecentlyPlayed(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		http.Error(w, `{"message":"Unauthorized"}`, http.StatusUnauthorized)
		return
	}

	ctx, cancel := context.WithTimeout(r.Context(), 5*time.Second)
	defer cancel()

	querySQL := `
		SELECT game_id, game_name, last_played
		FROM recently_played
		WHERE user_id = $1
		ORDER BY last_played DESC
		LIMIT 6;
	`

	rows, err := h.pool.Query(ctx, querySQL, userID)
	if err != nil {
		http.Error(w, `{"message":"Server error loading recently played"}`, http.StatusInternalServerError)
		return
	}
	defer rows.Close()

	list := make([]models.RecentGame, 0)
	for rows.Next() {
		var gameID, gameName string
		var lastPlayed time.Time

		if err := rows.Scan(&gameID, &gameName, &lastPlayed); err == nil {
			list = append(list, models.RecentGame{
				ID:         gameID,
				GameID:     gameID,
				Title:      gameName,
				GameName:   gameName,
				LastPlayed: lastPlayed.Format(time.RFC3339),
			})
		}
	}

	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(list)
}

// TrackRecentlyPlayed records a play session
func (h *RecentlyPlayedHandler) TrackRecentlyPlayed(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		http.Error(w, `{"message":"Unauthorized"}`, http.StatusUnauthorized)
		return
	}

	var req models.TrackGameRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, `{"message":"Invalid request body"}`, http.StatusBadRequest)
		return
	}

	if req.GameID == "" {
		http.Error(w, `{"message":"Game ID is required"}`, http.StatusBadRequest)
		return
	}

	gameName := req.GameName
	if gameName == "" {
		gameName = req.GameID
	}

	ctx, cancel := context.WithTimeout(r.Context(), 5*time.Second)
	defer cancel()

	upsertSQL := `
		INSERT INTO recently_played (user_id, game_id, game_name, last_played)
		VALUES ($1, $2, $3, NOW())
		ON CONFLICT (user_id, game_id)
		DO UPDATE SET last_played = NOW(), game_name = EXCLUDED.game_name;
	`
	_, err := h.pool.Exec(ctx, upsertSQL, userID, req.GameID, gameName)
	if err != nil {
		http.Error(w, `{"message":"Server error saving recently played"}`, http.StatusInternalServerError)
		return
	}

	// Return updated list
	h.GetRecentlyPlayed(w, r)
}
